### PR TITLE
Use of exponentialTimeout instead of ImmediateTimeout for resource scaling

### PR DESCRIPTION
**Pull request by dipankardas011**
#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
As its a TODO and we need to make the wait for certain interval as exponential timeout rather than linear timeout

#### Which issue(s) this PR fixes:

https://github.com/kubernetes/kubernetes/blob/c3e7eca7fd38454200819b60e58144d5727f1bbc/test/utils/update_resources.go#L31
#### Special notes for your reviewer:
Not sure how what all things need to be changed

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by dipankardas011** [comment]
Just not so clear about what change needs to be their and how we should proceed 😕

**Comment by dipankardas011** [comment]
Can anyone tell me what should be the best place to add the exponential logic?
also, exponential means 2^n?

**Comment by dipankardas011** [comment]
# questions
## does this line of does has timeOut for the resources getting changes i.e. its manifest inside the cluster gets changed?

https://github.com/kubernetes/kubernetes/blob/7ad8303b9602252b18932d34377cd2793595d537/test/utils/update_resources.go#LL53C22-L53C22
```go
    err := wait.PollImmediate(updateRetryInterval, updateRetryTimeout, cond)
```

**Comment by dipankardas011** [comment]
> # questions
> ## does this line of does has timeOut for the resources getting changes i.e. its manifest inside the cluster gets changed?
> https://github.com/kubernetes/kubernetes/blob/7ad8303b9602252b18932d34377cd2793595d537/test/utils/update_resources.go#L53
> 
> ## and does this part of code deals with the time new replicas become active by the changed configuration of a particular resources?
> https://github.com/kubernetes/kubernetes/blob/7ad8303b9602252b18932d34377cd2793595d537/test/utils/update_resources.go#L55

@cici37 @apelisse @johnSchnake

**Comment by dipankardas011** [comment]
currently planning how to go about this PR 😄

**Comment by dipankardas011** [comment]
https://github.com/kubernetes/kubernetes/blob/7ad8303b9602252b18932d34377cd2793595d537/staging/src/k8s.io/kubectl/pkg/scale/scale.go#L190-L194

cannot understand this piece of code 😞

**Comment by dipankardas011** [comment]
I am no longer interested.

**Comment by dinhxuanvu** [comment]
@dipankardas011 Hi there. You can open a PR with your implementation and then someone can review the code and go from there. Or you can ask questions on sig-api-machinery or sig-cli channels on Kubernetes Slack.

**Comment by dipankardas011** [comment]
> @dipankardas011 Hi there. You can open a PR with your implementation and then someone can review the code and go from there. Or you can ask questions on sig-api-machinery or sig-cli channels on Kubernetes Slack.

😃

File: test/utils/update_resources.go
Comment (dinhxuanvu) [inline test/utils/update_resources.go:36]: Can you explain why the retry number is set to 12? Just arbitrary or there is a meaning behind this?

File: test/utils/update_resources.go
Comment (dinhxuanvu) [inline test/utils/update_resources.go:60]: Should remove these two empty lines. Not needed.

File: test/utils/update_resources.go
Comment (dinhxuanvu) [inline test/utils/update_resources.go:52]: Maybe add a comment to tell what would be the backoff time at 1st, 6th and 12th attempts.

**Review by dinhxuanvu — CHANGES_REQUESTED** [review]

File: test/utils/update_resources.go
Comment (dipankardas011) [reply test/utils/update_resources.go:36]: It's is random one now sure how many time to retry

File: test/utils/update_resources.go
Comment (dipankardas011) [reply test/utils/update_resources.go:52]: Sure

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: Currently the No of retries are 12 which is the Steps
Duration is the initial duration = 5s
Factor decides by how much times the duration of that current iteration increases here taken 2x
also here specified the Jitter it adds a small amount of balancing factor to this exponential growth to eliminate the race conditions

for 
iter 0 -> 5s
iter 1 -> 10s
iter 2 -> 20s
iter 3 -> 40s
...

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: should we instead use the other factor or its looks good?
Currently
5 ( 1 + 2 + 2^2 + ... 2^(n-1)

File: test/utils/update_resources.go
Comment (dinhxuanvu) [inline test/utils/update_resources.go:35]: Nit: type `retries`

**Review by dinhxuanvu — COMMENTED** [review]
I think the code looks fine overall. Now, we do need an owner/expert on this component to review and approve the PR. I don't know who to tag for this. Probably need to go through commit history or owner file to find one.

**Comment by dipankardas011** [comment]
> 

sure

**Comment by dipankardas011** [comment]
Can anyone of you review this PR
Sorry for the ping
@shyamjvs
@krzysied
@seans3
@knight42

File: test/utils/update_resources.go
Comment (minherz) [inline test/utils/update_resources.go:52]: nit: should it be decorated with `const`?

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: Not sure what do you mean?

Comment (minherz) [inline test/utils/update_resources.go:52]:
```suggestion
	const exponentialBackoff := wait.Backoff{
```

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: It looks like we can be how to test things in this directory?

Comment (minherz) [reply test/utils/update_resources.go:52]: I am unsure what do you ask. `exponentialBackoff` object is no used anywhere except in the call to `wait.ExponentialBackoff()`. IMHO it is a good practice to use `const` to emphasis that this variable is not intended for a change.
Anyway, it was proposed as a suggestion.  AFAIK it should not block you.

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: Sure will add that

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: @minherz can you LGTM and approve this pr?

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: Once I have added the const

Comment (dipankardas011) [reply test/utils/update_resources.go:52]: image

Comment (minherz) [reply test/utils/update_resources.go:52]: > can you LGTM and approve this pr?

I do not have permissions to LGTM it :slightly_frowning_face: 

Regarding my comment about `const`, I apologies for confusing you. It was my honest mistake. Go does not support constant structure types.

**Review by minherz — APPROVED** [review]

File: test/utils/update_resources.go
Comment (dipankardas011) [reply test/utils/update_resources.go:52]: No issues 👍

**Comment by dipankardas011** [comment]
@minherz do you know anyone who can LGTM and approve it?

**Comment by dipankardas011** [comment]
/@wojtek-t can you take a look on this PR 😇
Sorry for directly tagging you

**Comment by aojea** [comment]
I don't feel this TODO is accurate, this is only used in a test
```
grep -r ScaleResourceWithRetries 
test/utils/update_resources.go:func ScaleResourceWithRetries(scalesGetter scaleclient.ScalesGetter, namespace, name string, size uint, gvr schema.GroupVersionResource) error {
test/e2e/framework/resource/resources.go:       if err := testutils.ScaleResourceWithRetries(scalesGetter, ns, name, size, gvr); err != nil {
```

and I do not  see the benefit of exponential backoff for asserting on tests , it make the test to take more time

**Comment by dipankardas011** [comment]
> I don't feel this TODO is accurate, this is only used in a test
> 
> ```
> grep -r ScaleResourceWithRetries 
> test/utils/update_resources.go:func ScaleResourceWithRetries(scalesGetter scaleclient.ScalesGetter, namespace, name string, size uint, gvr schema.GroupVersionResource) error {
> test/e2e/framework/resource/resources.go:       if err := testutils.ScaleResourceWithRetries(scalesGetter, ns, name, size, gvr); err != nil {
> ```
> 
> and I do not see the benefit of exponential backoff for asserting on tests , it make the test to take more time

I'm wondering what your call should I remove the code changes and the TODO comment?

**Comment by dipankardas011** [comment]
please anyone can review this PR?

**Review by minherz — APPROVED** [review]

**Pull request closed without merging**
